Terraform crashes when you pass null value to availability zone - azure

<!---
Please note the following potential times when an issue might be in Terraform core:
When  you missed to pass value to availability zone, it crashes. 
* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform (and AzureRM Provider) Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* `azurerm_2.24`
*`azurerm_virtual_machine`

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output
azurerm_virtual_machine.main: Creating...\n\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\nTerraform crashed! This is always indicative of a bug within Terraform.\nA crash log has been placed at \"crash.log\" relative to your current\nworking directory. It would be immensely helpful if you could please\nreport the crash with Terraform[1] so that we can fix this.\n\nWhen reporting bugs, please include your terraform version. That\ninformation is available on the first line of crash.log. You can also\nget it by running 'terraform --version' on the command line.\n\nSECURITY WARNING: the \"crash.log\" file that was created may contain \nsensitive information that must be redacted before it is safe to share \non the issue tracker.\n\n[1]: https://github.com/hashicorp/terraform/issues\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\nstderr: \nError: rpc error: code = Unavailable desc = transport is closing\n\n\npanic: interface conversion: interface {} is nil, not string\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: goroutine 211 [running]:\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure.ExpandZones(0xc000678610, 0x1, 0x1, 0x3eb10c0)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure/zones.go:54 +0x1e4\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute.resourceArmVirtualMachineCreateUpdate(0xc000cc7490, 0x4002a60, 0xc000300280, 0x0, 0x0)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute/virtual_machine_resource.go:654 +0x338\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc000236480, 0xc00017e910, 0xc00071cdc0, 0x4002a60, 0xc000300280, 0x4029e01, 0xc0013890b8, 0xc0014181e0)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:310 +0x365\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc00061e700, 0xc001353a10, 0xc00017e910, 0xc00071cdc0, 0xc001389118, 0xc00000f708, 0x402c040)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/provider.go:294 +0x99\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc000617008, 0x4ea8020, 0xc0013bdaa0, 0xc000cc6e70, 0xc000617008, 0xc0013bdaa0, 0xc001213b78)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:885 +0x8b4\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x44051c0, 0xc000617008, 0x4ea8020, 0xc0013bdaa0, 0xc0000d9260, 0x0, 0x4ea8020, 0xc0013bdaa0, 0xc000517400, 0xb45)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5/tfplugin5.pb.go:3305 +0x217\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0005c2480, 0x4ee5780, 0xc0005c2d80, 0xc000da8800, 0xc0007d4330, 0x7d2bc80, 0x0, 0x0, 0x0)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1024 +0x501\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: google.golang.org/grpc.(*Server).handleStream(0xc0005c2480, 0x4ee5780, 0xc0005c2d80, 0xc000da8800, 0x0)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1313 +0xd3d\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000ad36e0, 0xc0005c2480, 0x4ee5780, 0xc0005c2d80, 0xc000da8800)\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:722 +0xa1\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: created by google.golang.org/grpc.(*Server).serveStreams.func1\n2020-08-24T13:01:54.678Z [DEBUG] plugin.terraform-provider-azurerm_v2.24.0_x5: \t/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:720 +0xa1\n2020-08-24T13:01:54.681Z [DEBUG] plugin: plugin process exited: path=/tmp/awx_1462_ge90inqc/project/Playbooks/IaaS_VM/.terraform/plugins/linux_amd64/terraform-provider-azurerm_v2.24.0_x5 pid=923 error=\"exit status 2\"\n2020/08/24 13:01:54 [DEBUG] azurerm_virtual_machine.main: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalMaybeTainted\n2020/08/24 13:01:54 [TRACE] EvalMaybeTainted: azurerm_virtual_machine.main encountered an error during creation, so it is now marked as tainted\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalWriteState\n2020/08/24 13:01:54 [TRACE] EvalWriteState: removing state object for azurerm_virtual_machine.main\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalApplyProvisioners\n2020/08/24 13:01:54 [TRACE] EvalApplyProvisioners: azurerm_virtual_machine.main has no state, so skipping provisioners\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalMaybeTainted\n2020/08/24 13:01:54 [TRACE] EvalMaybeTainted: azurerm_virtual_machine.main encountered an error during creation, so it is now marked as tainted\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalWriteState\n2020/08/24 13:01:54 [TRACE] EvalWriteState: removing state object for azurerm_virtual_machine.main\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalIf\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalIf\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalWriteDiff\n2020/08/24 13:01:54 [TRACE] <root>: eval: *terraform.EvalApplyPost\n2020/08/24 13:01:54 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing\n2020/08/24 13:01:54 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing\n2020/08/24 13:01:54 [TRACE] [walkApply] Exiting eval tree: azurerm_virtual_machine.main\n2020/08/24 13:01:54 [TRACE] vertex \"azurerm_virtual_machine.main\": visit complete\n2020/08/24 13:01:54 [TRACE] dag/walk: upstream of \"azurerm_backup_protected_vm.main\" errored, so skipping\n2020/08/24 13:01:54 [TRACE] dag/walk: upstream of \"meta.count-boundary (EachMode fixup)\" errored, so skipping\n2020/08/24 13:01:54 [TRACE] dag/walk: upstream of \"provider.azurerm (close)\" errored, so skipping\n2020/08/24 13:01:54 [TRACE] dag/walk: upstream of \"root\" errored, so skipping\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: reading latest snapshot from terraform.tfstate\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: read snapshot with lineage \"fd28c42c-682e-7407-de9d-03a8f10e8383\" serial 1\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: no original state snapshot to back up\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 2\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info\n2020/08/24 13:01:54 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock\n2020-08-24T13:01:54.688Z [DEBUG] plugin: plugin exited\n\n"
<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

<!--- What should have happened? --->

### Actual Behavior

<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in a Azure China/Germany/Government? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

* #0000

--- a/azurerm/helpers/azure/zones.go
+++ b/azurerm/helpers/azure/zones.go
@@ -1,6 +1,9 @@
 package azure
 
-import "github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azuread/azuread/helpers/validate"
+)
 
 func SchemaZones() *schema.Schema {
 	return &schema.Schema{
@@ -9,6 +12,7 @@ func SchemaZones() *schema.Schema {
 		ForceNew: true,
 		Elem: &schema.Schema{
 			Type: schema.TypeString,
+			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
 }
@@ -21,6 +25,7 @@ func SchemaSingleZone() *schema.Schema {
 		MaxItems: 1,
 		Elem: &schema.Schema{
 			Type: schema.TypeString,
+			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
 }
@@ -33,6 +38,7 @@ func SchemaMultipleZones() *schema.Schema {
 		MinItems: 1,
 		Elem: &schema.Schema{
 			Type: schema.TypeString,
+			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
 }
@@ -44,6 +50,7 @@ func SchemaZonesComputed() *schema.Schema {
 		Computed: true,
 		Elem: &schema.Schema{
 			Type: schema.TypeString,
+			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
 }

--- a/azurerm/helpers/azure/zones.go
+++ b/azurerm/helpers/azure/zones.go
@@ -11,7 +11,7 @@ func SchemaZones() *schema.Schema {
 		Optional: true,
 		ForceNew: true,
 		Elem: &schema.Schema{
-			Type: schema.TypeString,
+			Type:         schema.TypeString,
 			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
@@ -24,7 +24,7 @@ func SchemaSingleZone() *schema.Schema {
 		ForceNew: true,
 		MaxItems: 1,
 		Elem: &schema.Schema{
-			Type: schema.TypeString,
+			Type:         schema.TypeString,
 			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
@@ -37,7 +37,7 @@ func SchemaMultipleZones() *schema.Schema {
 		ForceNew: true,
 		MinItems: 1,
 		Elem: &schema.Schema{
-			Type: schema.TypeString,
+			Type:         schema.TypeString,
 			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}
@@ -49,7 +49,7 @@ func SchemaZonesComputed() *schema.Schema {
 		Optional: true,
 		Computed: true,
 		Elem: &schema.Schema{
-			Type: schema.TypeString,
+			Type:         schema.TypeString,
 			ValidateFunc: validate.NoEmptyStrings,
 		},
 	}

--- a/azurerm/helpers/azure/zones.go
+++ b/azurerm/helpers/azure/zones.go
@@ -2,7 +2,7 @@ package azure
 
 import (
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
-	"github.com/terraform-providers/terraform-provider-azuread/azuread/helpers/validate"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
 )
 
 func SchemaZones() *schema.Schema {
@@ -12,7 +12,7 @@ func SchemaZones() *schema.Schema {
 		ForceNew: true,
 		Elem: &schema.Schema{
 			Type:         schema.TypeString,
-			ValidateFunc: validate.NoEmptyStrings,
+			ValidateFunc: validation.StringIsNotEmpty,
 		},
 	}
 }
@@ -25,7 +25,7 @@ func SchemaSingleZone() *schema.Schema {
 		MaxItems: 1,
 		Elem: &schema.Schema{
 			Type:         schema.TypeString,
-			ValidateFunc: validate.NoEmptyStrings,
+			ValidateFunc: validation.StringIsNotEmpty,
 		},
 	}
 }
@@ -38,7 +38,7 @@ func SchemaMultipleZones() *schema.Schema {
 		MinItems: 1,
 		Elem: &schema.Schema{
 			Type:         schema.TypeString,
-			ValidateFunc: validate.NoEmptyStrings,
+			ValidateFunc: validation.StringIsNotEmpty,
 		},
 	}
 }
@@ -50,7 +50,7 @@ func SchemaZonesComputed() *schema.Schema {
 		Computed: true,
 		Elem: &schema.Schema{
 			Type:         schema.TypeString,
-			ValidateFunc: validate.NoEmptyStrings,
+			ValidateFunc: validation.StringIsNotEmpty,
 		},
 	}
 }

